exporters/metric/dogstatsd/internal/statsd/conn.go

```diff
@@ -0,0 +1,310 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statsd
+
+// See https://github.com/b/statsd_spec for the best-available statsd
+// syntax specification.  See also
+// https://github.com/statsd/statsd/edit/master/docs/metric_types.md
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"strconv"
+
+	"go.opentelemetry.io/otel/api/core"
+	"go.opentelemetry.io/otel/api/unit"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+)
+
+type (
+	// Config supports common configuration that applies to statsd exporters.
+	Config struct {
+		// URL describes the destination for exporting statsd data.
+		// e.g., udp://host:port
+		//       tcp://host:port
+		//       unix:///socket/path
+		URL string
+
+		// Writer is an alternate to providing a URL.  When Writer is
+		// non-nil, URL will be ignored and the exporter will write to
+		// the configured Writer interface.
+		Writer io.Writer
+
+		// MaxPacketSize this limits the packet size for packet-oriented transports.
+		MaxPacketSize int
+
+		// TODO support Dial and Write timeouts
+	}
+
+	// Exporter is common type meant to implement concrete statsd
+	// exporters.
+	Exporter struct {
+		adapter Adapter
+		config  Config
+		conn    net.Conn
+		writer  io.Writer
+		buffer  bytes.Buffer
+	}
+
+	// Adapter supports statsd syntax variations, primarily plain
+	// statsd vs. dogstatsd.
+	Adapter interface {
+		AppendName(export.Record, *bytes.Buffer)
+		AppendTags(export.Record, *bytes.Buffer)
+	}
+)
+
+const (
+	formatCounter   = "c"
+	formatHistogram = "h"
+	formatGauge     = "g"
+	formatTiming    = "ms"
+
+	// MaxPacketSize defaults to Ethernet Jumbo frame size.  If
+	// this is too large, you will see "write: message too long"
+	// errors.
+	MaxPacketSize = 9216
```

Comment:
Jumbo frames don't really exist in a any consistent form. AWS, for example, has a jumbo frame features that caps at 9001 bytes per packet. Most cloud providers stick to the 1500 MTU standard. I'd encourage a default size that is less than 1500 in order to account for both the required and optional vendor headers in the packets.

For reference, the official datadog client for Go uses 1432 (https://github.com/DataDog/datadog-go/blob/f137255bc8cca1abb54c897418b6bb752acd5839/statsd/statsd.go#L37).

Reply:
We'll use this value. Thanks.